### ISSUE TITLE
Use dynamic strict CSP policy

### ISSUE THREAD

**Issue by ibennetch** [phpmyadmin/phpmyadmin#12357]
4) use Dynamic Strict CSP Policy
 
There are 2 open issues about unsafe-inline: #12262 #12261
 
currently the CSP has unsafe-eval directive, it's recommended to avoid all calls that this directive allow, and then remove unsafe-eval directive.
 
also it's recommended to use dynamic CSP policy, for example the *.tile.openstreetmap.org for img-src is always added.
it's better to add it only when needed (for example: not in login page)

**Comment by ibennetch** [comment]
Suggested by Emanuel Bronshtein

**Comment by emanuelb** [comment]
In addition the $captcha_url
[https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L529](https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/Header.php#L529)
is needed only when login is displayed, at:
[https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/plugins/auth/AuthenticationCookie.php#L203](https://github.com/phpmyadmin/phpmyadmin/blob/master/libraries/plugins/auth/AuthenticationCookie.php#L203)
thus it's better to set $captcha_url to '' in case the login page will not be returned.

**Comment by nijel** [comment]
I don't think this would really work with AJAX - you define CSP on initial page load, but you end up needing additional permission on later AJAX request and that's simply not possible. There would be option to reload whole page on this event, but I don't think it's worth of the troubles.

So the only practical differentiation is to use separate CSP for login page and for application itself.

**Comment by emanuelb** [comment]
yeah, it depend on how navigation is implemented, still there are scenarios when it's known in advance that some directive is needed/not-needed (for example: captcha related stuff)
anyway, it's usually a lot easier to reason about & make the work toward dynamic CSP when most of the application is already play nice with it, which mean starting with fixing/closing issues #12262  #13270 #12782
